examples/vision/aodnet.py

```diff
@@ -0,0 +1,376 @@
+Last modified: 2021/04/10
+Description: Train a lightweight Deep CNN for image denhazing.
+"""
+"""
+# Introduction
+
+Image Dehazing is an important problem is Computer Vision. It can not only be used to
+enhance photographs clicked in hazy conditions, but also can be used to augment other
+computer vision systems such as Object Detection and Seantic Segmentation as a
+pre-processing step. In this example, we implement
+[AODNet](https://arxiv.org/abs/1707.06543v1), a simple and light-weight Deep CNN model
+that directly generates a clean image from a hazy image. The end-to-end design of AODNet
+makes it easy to embed AOD-Net into other deep models such as Faster-RCNN and YOLO as a
+pre-processing step, making it easier for such model to detect and recognize objects from
+a clean image.
+"""
+
+"""shell
+!nvidia-smi -L
+"""
+
+import os
+import gdown
+import numpy as np
+from PIL import Image
+from glob import glob
+from glob import glob
+import tensorflow as tf
+import matplotlib.pyplot as plt
+
+"""
+# Downloading Dataset
+
+The authors of the AODNet paper have created synthesized hazy images using the
+groundtruth images with depth meta-data from the indoor [NYU2
+Depth Database](https://cs.nyu.edu/~silberman/datasets/nyu_depth_v2.html). We will
+download this dataset from Google Drive using `gdown`.
+"""
+
+gdown.download(
+    "https://drive.google.com/uc?id=1sInD9Ydq8-x7WwqehE0EyRknMdSFPOat",
+    "Dehaze-NYU.zip",
+    quiet=False,
+)
+
+os.system("unzip -qq Dehaze-NYU.zip")
+os.system("rm ./Dehaze-NYU.zip")
+
+"""
+## Data Loader
+
+In order to ensure that the model is fed the data efficiently during training, we will be
+use the `tf.data` API to create our data loading pipeline. Due to memory constraints for
+training on Google Colab, we generate random crops of size `256` from corresponding
+hazing and clean image pair. For the training dataset, we use also apply random rotation
+and random horizontal flips in order to augment the data.
+"""
+
+
+def get_image_file_list(dataset_path):
+    original_image_files = []
+    hazy_image_paths = sorted(
+        glob(str(os.path.join(dataset_path, "train_images/*.jpg")))
+    )
+    for image_path in hazy_image_paths:
+        image_file_name = image_path.split("/")[-1]
+        original_file_name = (
+            image_file_name.split("_")[0] + "_" + image_file_name.split("_")[1] + ".jpg"
+        )
+        original_image_files.append(
+            str(os.path.join(dataset_path, "original_images/" + original_file_name))
+        )
+    return original_image_files, hazy_image_paths
+
+
+def read_images(image_files):
+    dataset = tf.data.Dataset.from_tensor_slices(image_files)
+    dataset = dataset.map(tf.io.read_file)
+    dataset = dataset.map(
+        lambda x: tf.image.decode_png(x, channels=3),
+        num_parallel_calls=tf.data.experimental.AUTOTUNE,
+    )
+    return dataset
+
+
+def random_crop(hazy_image, original_image, low_crop_size, enhanced_crop_size):
```

Comment:
Please use preprocessing layers for data augmentation, like in this example: https://keras.io/examples/vision/image_classification_from_scratch/#using-image-data-augmentation

You can do this with the `RandomCrop` layer.